File: array_sort/sort_quick/src/QuickSortDemo.py
# Python
# 配列の並び替え: クイックソート (Quick Sort)

class ArrayData:
    def __init__(self):
        self._data = []

    def get(self):
        return self._data

    def set(self, data):
        self._data = data
        return True

    def _quick_sort(self, target):
        # 空の配列または要素が1つの場合はそのまま返す（基底条件）
        if len(target) <= 1:
            return target
        
        # ピボットを選択（この実装では最後の要素を選択）
        pivot = target[-1]
        
        # ピボットより小さい要素と大きい要素に分ける
        left = []
        right = []
        
        # 最後の要素（ピボット）を除いて配列をスキャン
        for i in range(len(target) - 1):
            if target[i] <= pivot:
                left.append(target[i])
            else:
                right.append(target[i])
        
        # 左側の部分配列、ピボット、右側の部分配列を再帰的にソートして結合
        return self._quick_sort(left) + [pivot] + self._quick_sort(right)

    def sort(self):
        self._quick_sort(self._data)
        return True

def main():
    print("QuickSort TEST -----> start")

    array_data = ArrayData()

    # ランダムな整数の配列
    print(f"\nsort")
    input = [64, 34, 25, 12, 22, 11, 90]
    print(f"  ソート前: {input}")
    array_data.set(input)
    array_data.sort()
    print(f"  ソート後: {array_data.get()}")
    
    # 既にソートされている配列
    print(f"\nsort")
    input = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    print(f"  ソート前: {input}")
    array_data.set(input)
    array_data.sort()
    print(f"  ソート後: {array_data.get()}")
    
    # 逆順の配列
    print(f"\nsort")
    input = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
    print(f"  ソート前: {input}")
    array_data.set(input)
    array_data.sort()
    print(f"  ソート後: {array_data.get()}")
    
    # 重複要素を含む配列
    print(f"\nsort")
    input = [10, 9, 8, 7, 6, 10, 9, 8, 7, 6]
    print(f"  ソート前: {input}")
    array_data.set(input)
    array_data.sort()
    print(f"  ソート後: {array_data.get()}")
    
    # 空の配列
    print(f"\nsort")
    input = []
    print(f"  ソート前: {input}")
    array_data.set(input)
    array_data.sort()
    print(f"  ソート後: {array_data.get()}")

    print("\nQuickSort TEST <----- end")

if __name__ == "__main__":
    main()

File: array_sort/sort_quick/src/QuickSortDemo.java
// Java
// 配列の並び替え: クイックソート (Quick Sort)

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class ArrayData {
    private List<Integer> data;

    public ArrayData() {
        this.data = new ArrayList<>();
    }

    public List<Integer> get() {
        return this.data;
    }

    public boolean set(List<Integer> data) {
        this.data = new ArrayList<>(data);
        return true;
    }

    private List<Integer> quickSort(List<Integer> target) {
        // 空の配列または要素が1つの場合はそのまま返す（基底条件）
        if (target.size() <= 1) {
            return target;
        }
        
        // ピボットを選択（この実装では最後の要素を選択）
        Integer pivot = target.get(target.size() - 1);
        
        // ピボットより小さい要素と大きい要素に分ける
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();
        
        // 最後の要素（ピボット）を除いて配列をスキャン
        for (int i = 0; i < target.size() - 1; i++) {
            if (target.get(i) <= pivot) {
                left.add(target.get(i));
            } else {
                right.add(target.get(i));
            }
        }
        
        // 左側の部分配列、ピボット、右側の部分配列を再帰的にソートして結合
        List<Integer> result = new ArrayList<>();
        result.addAll(quickSort(left));
        result.add(pivot);
        result.addAll(quickSort(right));
        
        return result;
    }

    public boolean sort() {
        this.data = quickSort(this.data);
        return true;
    }
}

public class QuickSortDemo {
    public static void main(String[] args) {
        System.out.println("QuickSort TEST -----> start");

        ArrayData arrayData = new ArrayData();

        // ランダムな整数の配列
        System.out.println("\nsort");
        List<Integer> input = Arrays.asList(64, 34, 25, 12, 22, 11, 90);
        System.out.println("  ソート前: " + input);
        arrayData.set(input);
        arrayData.sort();
        System.out.println("  ソート後: " + arrayData.get());
        
        // 既にソートされている配列
        System.out.println("\nsort");
        input = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
        System.out.println("  ソート前: " + input);
        arrayData.set(input);
        arrayData.sort();
        System.out.println("  ソート後: " + arrayData.get());
        
        // 逆順の配列
        System.out.println("\nsort");
        input = Arrays.asList(10, 9, 8, 7, 6, 5, 4, 3, 2, 1);
        System.out.println("  ソート前: " + input);
        arrayData.set(input);
        arrayData.sort();
        System.out.println("  ソート後: " + arrayData.get());
        
        // 重複要素を含む配列
        System.out.println("\nsort");
        input = Arrays.asList(10, 9, 8, 7, 6, 10, 9, 8, 7, 6);
        System.out.println("  ソート前: " + input);
        arrayData.set(input);
        arrayData.sort();
        System.out.println("  ソート後: " + arrayData.get());
        
        // 空の配列
        System.out.println("\nsort");
        input = new ArrayList<>();
        System.out.println("  ソート前: " + input);
        arrayData.set(input);
        arrayData.sort();
        System.out.println("  ソート後: " + arrayData.get());

        System.out.println("\nQuickSort TEST <----- end");
    }
}